config/sources/common.conf

```diff
@@ -6,19 +6,180 @@
+	else
+		display_alert "LINUXFAMILY is set by board/family/hooks to '${LINUXFAMILY}'" "common_defaults_for_mainline" "ext"
+	fi
+
+	if [[ -z ${KERNELPATCHDIR} ]]; then
+		display_alert "KERNELPATCHDIR is unset" "common_defaults_for_mainline" "info"
+	else
+		display_alert "KERNELPATCHDIR is set by board/family/hooks to '${KERNELPATCHDIR}'" "common_defaults_for_mainline" "ext"
+	fi
+
+	if [[ -z ${KERNEL_PATCH_ARCHIVE_BASE} ]]; then
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is unset" "common_defaults_for_mainline" "info"
+	else
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is set by board/family/hooks to '${KERNELPATCHDIR}'" "common_defaults_for_mainline" "ext"
+	fi
+
+	if [[ -z ${LINUXCONFIG} ]]; then
+		display_alert "LINUXCONFIG is unset" "common_defaults_for_mainline" "info"
+	else
+		display_alert "LINUXCONFIG is set by board/family/hooks to '${LINUXCONFIG}'" "common_defaults_for_mainline" "ext"
+	fi
+
+	# Generic defaultings, used for all kernels (including vendor/legacy/3rd-party and mainline).
+	# If KERNEL_PATCH_ARCHIVE_BASE is unset, set it to LINUXFAMILY.
+	if [[ -z ${KERNEL_PATCH_ARCHIVE_BASE} ]]; then
+		display_alert "KERNEL_PATCH_ARCHIVE_BASE is unset; using LINUXFAMILY: '${LINUXFAMILY}'" "common_defaults_for_mainline" "info"
+		KERNEL_PATCH_ARCHIVE_BASE="$LINUXFAMILY"
+	fi
+
+	# If KERNELPATCHDIR is unset, set it to "archive/${KERNELPATCHDIR}-${KERNEL_MAJOR_MINOR}"
+	if [[ -z ${KERNELPATCHDIR} ]]; then
+		display_alert "KERNELPATCHDIR is unset; using 'archive/${KERNEL_PATCH_ARCHIVE_BASE}-${KERNEL_MAJOR_MINOR}'" "common_defaults_for_mainline" "info"
+		KERNELPATCHDIR="archive/${KERNEL_PATCH_ARCHIVE_BASE}-${KERNEL_MAJOR_MINOR}" # previously was KERNELPATCHDIR="$LINUXFAMILY-$BRANCH"
+	fi
+
+	# If LINUFAMILY is unset... bomb
+	if [[ -z ${LINUXFAMILY} ]]; then
+		exit_with_error "LINUXFAMILY is unset after family config; KERNELSOURCE=${KERNELSOURCE}"
+	fi
+
+	# if LINUXCONFIG is unset... default to linux-${LINUXFAMILY}-${BRANCH} # Attention: no KERNEL_MAJOR_MINOR here, so manual file rollover is necessary
+	if [[ -z ${LINUXCONFIG} ]]; then
+		display_alert "LINUXCONFIG is unset; using 'linux-${LINUXFAMILY}-${BRANCH}'" "common_defaults_for_mainline" "info"
+		LINUXCONFIG="linux-${LINUXFAMILY}-${BRANCH}"
+	fi
+
+	# Mainline defaults
+
+	if [[ -z ${KERNELSOURCE} ]]; then
+		display_alert "Using mainline kernel defaults" "common_defaults_for_mainline: KERNEL_MAJOR_MINOR: ${KERNEL_MAJOR_MINOR}" "info"
+
+		# this "manifests"/resolves the mirrors; any changes after this will need to resolve their own mirrors; later hooks won't be subject to mirroring
+		KERNELSOURCE="${MAINLINE_KERNEL_SOURCE}"
+
+		# If the branch is unset, auto-determine it based on KERNEL_MAJOR_MINOR; allow hooks to customize
+		if [[ -z ${KERNELBRANCH} ]]; then
+			declare original_kernelsource="${KERNELSOURCE}" # store for later comparing
+			call_extension_method "mainline_kernel_decide_version" <<- 'MAINLINE_KERNEL_DECIDE_VERSION'
```

Comment:
I am not sure if this should be after all the checks you are adding. Someone can end up using this to unset all the variables that are being set. For example whats stopping me from putting `unset LINUXCONFIG` and `unset LINUXFAMILY` into this extension call.

Reply:
Also I see you are not adding a hook function for the uboot equivalent of this function. So why do we need it here?

Reply:
None of this has the pretense of being resistant to such unset's. None of Armbian is. Users can use lib.config to unset the hell out of everything anyway.

Ref u-boot: I have moved the minimal set of pre-settage that was done in common to a hook. U-boot does not entail the same level of "auto-updating" that kernel does (eg receives no point releases, only major.minor) so it's actually much simpler.